pkg/vault/contracts/Vault.sol

```diff
@@ -661,17 +665,29 @@ contract Vault is IVaultMain, VaultCommon, Proxy {
             revert BptAmountOutBelowMin(bptAmountOut, params.minBptAmountOut);
         }
 
+        _ensureValidTradeAmount(bptAmountOut);
+
         amountsInRaw = new uint256[](locals.numTokens);
 
         for (uint256 i = 0; i < locals.numTokens; ++i) {
+            uint256 amountInRaw;
+
             // 1) Calculate raw amount in.
-            // amountsInRaw are amounts actually entering the Pool, so we round up.
-            // Do not mutate in place yet, as we need them scaled for the `onAfterAddLiquidity` hook
-            uint256 amountInRaw = amountsInScaled18[i].toRawUndoRateRoundUp(
-                poolData.decimalScalingFactors[i],
-                poolData.tokenRates[i]
-            );
-            amountsInRaw[i] = amountInRaw;
+            {
+                uint256 amountInScaled18 = amountsInScaled18[i];
+                if (amountInScaled18 > 0) {
```

Comment:
Interesting.

We're already checking for this in `_ensureValidTradeAmount`, so this basically is an optimization for the math in `toRawUndoRateRoundUp`. What if we do it inside the library instead? The scope and the if inside a for makes the code a bit heavy to ready IMO. WDYT?

Reply:
Trying it. Have ScalingHelpers check for 0 for functions that do multiple operations (e.g., more than one multiply/divide). Not sure if the compiler already optimizes this?